README.md
# music use seajs

---

A nice music demo with seajs and spm.

---

## Usage

It is very easy to use this module.

````html
<div class="-sea-music">
</div>
````

```javascript
seajs.use('sea-music', function(seaMusic) {

});
```

## Api

Here is some details.


HISTORY.md
# History

---

## 1.0.0

`new` /sea-music First version.


tests/sea-music-spec.js
define(function(require) {

  var seaMusic = require('../src/sea-music');

  describe('sea-music', function() {

    it('normal usage', function() {

    });
  });

});


src/page/index/demo.js
$(document).ready(function () {
    //获取音频工具 
    var audio = document.getElementById("myMusic"); 


    //开始，暂停按钮
    $("#MainControl")._toggle(function () {
        $(this).removeClass("MainControl").addClass("StopControl");
        if (audio.src == "") {
            var Defaultsong = $(".Single .SongName").eq(0).attr("KV");
            $(".MusicBox .ProcessControl .SongName").text(Defaultsong);
            $(".Single .SongName").eq(0).css("color", "#7A8093");
            audio.src = "../Media/" + Defaultsong + ".mp3";
        }
        audio.play();
        TimeSpan();
    }, function () {
        $(this).removeClass("StopControl").addClass("MainControl");
        audio.pause();
    });


    //歌曲列表的选择操作
    $(".MusicList .List .Single .SongName").click(function () {
        $(".MusicList .List .Single .SongName").css("color", "#fff");
        $("#MainControl").removeClass("MainControl").addClass("StopControl");
        $(this).css("color", "#7A8093");
        var SongName = $(this).attr("KV");
        $(".MusicBox .ProcessControl .SongName").text(SongName);
        audio.src = "../Media/" + SongName + ".mp3";
        audio.play();
        TimeSpan();
    });

    //左前进按钮
    $(".LeftControl").click(function () {
        $(".MusicList .List .Single .SongName").each(function () {
            if ($(this).css("color") == "rgb(122, 128, 147)") {
                var IsTop = $(this).parent(".Single").prev(".Single").length == 0 ? true : false;  //检查是否是最顶端的歌曲
                var PrevSong;
                if (IsTop) {
                    PrevSong = $(".Single").last().children(".SongName").attr("KV");
                    $(".Single").last().children(".SongName").css("color", "#7A8093");
                }
                else {
                    PrevSong = $(this).parent(".Single").prev(".Single").children(".SongName").attr("KV");
                    $(this).parent(".Single").prev(".Single").children(".SongName").css("color", "#7A8093");
                }

                audio.src = "../Media/" + PrevSong + ".mp3";
                $(".MusicBox .ProcessControl .SongName").text(PrevSong);
                $(this).css("color", "#fff");
                audio.play();
                return false; //代表break
            }
        })
    });

    //右前进按钮
    $(".RightControl").click(function () {
        $(".MusicList .List .Single .SongName").each(function () {
            if ($(this).css("color") == "rgb(122, 128, 147)") {
                var IsBottom = $(this).parent(".Single").next(".Single").length == 0 ? true : false;  //检查是否是最尾端的歌曲
                var NextSong;
                if (IsBottom) {
                    NextSong = $(".Single").first().children(".SongName").attr("KV");
                    $(".Single").first().children(".SongName").css("color", "#7A8093");
                }
                else {
                    NextSong = $(this).parent(".Single").next(".Single").children(".SongName").attr("KV");
                    $(this).parent(".Single").next(".Single").children(".SongName").css("color", "#7A8093");
                }

                audio.src = "../Media/" + NextSong + ".mp3";
                $(".MusicBox .ProcessControl .SongName").text(NextSong);
                $(this).css("color", "#fff");
                audio.play();
                return false; //代表break
            }
        })
    });

    //静音按钮
    $(".VoiceEmp").click(function () {
        $(".VoidProcessYet").css("width", "0");
        audio.volume = 0;
    });

    //满音量按钮
    $(".VoiceFull").click(function () {
        $(".VoidProcessYet").css("width", "66px");
        audio.volume = 1;
    });

    /*
    之前一直考虑进度条的原理，这边进度条的做法启发自腾讯一款播放器的做法，采用两个2px高度的div，重叠，
    上面那个与下面那个div的颜色不一样，用于区分进度,顶层div的width是根据播放的长度来调整的，width越长，说明播放越长，
    知道上层的div完全覆盖下面的div，达到长度的一致，说明播放完毕。我们的播放进度条和音量进度条都是这样做的
    */

    // 音频进度条事件（进度增加）
    $(".Process").click(function (e) {

        //播放进度条的基准参数
        var Process = $(".Process").offset();
        var ProcessStart = Process.left;
        var ProcessLength = $(".Process").width();


        var CurrentProces = e.clientX - ProcessStart;
        DurationProcessRange(CurrentProces / ProcessLength);
        $(".ProcessYet").css("width", CurrentProces);
    });

    //音频进度条事件（进度减少）
    $(".ProcessYet").click(function (e) {

        //播放进度条的基准参数
        var Process = $(".Process").offset();
        var ProcessStart = Process.left;
        var ProcessLength = $(".Process").width();

        var CurrentProces = e.clientX - ProcessStart;
        DurationProcessRange(CurrentProces / ProcessLength);
        $(".ProcessYet").css("width", CurrentProces);
    });

    //音量进度条事件（进度增加）
    $(".VoidProcess").click(function (e) {
        //音量进度条的基准参数
        var VoidProcess = $(".VoidProcess").offset();
        var VoidProcessStart = VoidProcess.left;
        var VoidProcessLength = $(".VoidProcess").width();

        var CurrentProces = e.clientX - VoidProcessStart;
        VolumeProcessRange(CurrentProces / VoidProcessLength);
        $(".VoidProcessYet").css("width", CurrentProces);
    });

    //音量进度条时间（进度减少）
    $(".VoidProcessYet").click(function (e) {
        //音量进度条的基准参数
        var VoidProcess = $(".VoidProcess").offset();
        var VoidProcessStart = VoidProcess.left;
        var VoidProcessLength = $(".VoidProcess").width();

        var CurrentProces = e.clientX - VoidProcessStart;
        VolumeProcessRange(CurrentProces / VoidProcessLength);
        $(".VoidProcessYet").css("width", CurrentProces);
    });

    //显示音乐列表事件
    $(".ShowMusicList").toggle(function () {
        $(".MusicList").show();

        var MusicBoxRight = $(".MusicBox").offset().left + $(".MusicBox").width();
        var MusicBoxBottom = $(".MusicBox").offset().top + $(".MusicBox").height();
        $(".MusicList").css("left", MusicBoxRight - $(".MusicList").width());
        $(".MusicList").css("top", MusicBoxBottom + 15);
    }, function () {
        $(".MusicList").hide();
    });


    //监听媒体文件结束的事件（ended），这边一首歌曲结束就读取下一首歌曲，实现播放上的循环播放
    audio.addEventListener('ended', function () {
        $(".MusicList .List .Single .SongName").each(function () {
            if ($(this).css("color") == "rgb(122, 128, 147)") {
                var IsBottom = $(this).parent(".Single").next(".Single").length == 0 ? true : false;  //检查是否是最尾端的歌曲
                var NextSong;
                if (IsBottom) {
                    NextSong = $(".Single").first().children(".SongName").attr("KV");
                    $(".Single").first().children(".SongName").css("color", "#7A8093");
                }
                else {
                    NextSong = $(this).parent(".Single").next(".Single").children(".SongName").attr("KV");
                    $(this).parent(".Single").next(".Single").children(".SongName").css("color", "#7A8093");
                }

                audio.src = "../Media/" + NextSong + ".mp3";
                $(".MusicBox .ProcessControl .SongName").text(NextSong);
                $(this).css("color", "#fff");
                audio.play();
                return false; //代表break
            }
        });
    }, false);


});

//音量进度条的转变事件
function VolumeProcessRange(rangeVal) {
    var audio = document.getElementById("myMusic");
    audio.volume = parseFloat(rangeVal);
}

//播放进度条的转变事件
function DurationProcessRange(rangeVal) {
    var audio = document.getElementById("myMusic");
    audio.currentTime = rangeVal * audio.duration;
    audio.play();
}

//播放事件
function Play(obj) {
    var SongUrl = obj.getAttribute("SongUrl");
    var audio = document.getElementById("myMusic");
    audio.src = "../Media/" + SongUrl + ".mp3";
    audio.play();
    TimeSpan();
}

//暂停事件
function Pause() {
    var audio = document.getElementById("myMusic");
    $("#PauseTime").val(audio.currentTime);
    audio.pause();
}

//继续播放事件
function Continue() {
    var audio = document.getElementById("myMusic");
    audio.startTime = $("PauseTime").val();
    audio.play();
}

//时间进度处理程序
function TimeSpan() {
    var audio = document.getElementById("myMusic");
    var ProcessYet = 0;
    setInterval(function () {
        var ProcessYet = (audio.currentTime / audio.duration) * 500;
        $(".ProcessYet").css("width", ProcessYet);
        var currentTime = timeDispose(audio.currentTime);
        var timeAll = timeDispose(TimeAll());
        $(".SongTime").html(currentTime + "&nbsp;|&nbsp;" + timeAll);
    }, 1000);
}

//时间处理，因为时间是以为单位算的，所以这边执行格式处理一下
function timeDispose(number) {
    var minute = parseInt(number / 60);
    var second = parseInt(number % 60);
    minute = minute >= 10 ? minute : "0" + minute;
    second = second >= 10 ? second : "0" + second;
    return minute + ":" + second;
}

//当前歌曲的总时间
function TimeAll() {
    var audio = document.getElementById("myMusic");
    return audio.duration;
}

src/page/index/Audio.js
define(function(require, exports, module){
    var AudioFun = function(audio){
    		this.audio = audio;
    	},
    	proto = AudioFun.prototype;

    proto = {
    	//音量进度条的转变事件
    	VolumeProcessRange: function(rangeVal){
    		this.audio.volume = parseFloat(rangeVal);
    	},
    	//播放进度条的转变事件
    	DurationProcessRange: function(rangeVal){
    		this.audio.currentTime = rangeVal * audio.duration;
    		this.audio.play();
    	},
    	//播放事件
    	Play: function(SongUrl, callback){
    		this.audio.src = SongUrl;
		    this.audio.play();
		    callback && callback();
    	},
    	//暂停事件
    	Pause: function(callback){
    		this.audio.pause();
    		callback && callback(this.audio.currentTime);
    	},
    	//时间格式处理
    	TimeDispose: function(number){
    		var minute = parseInt(number / 60),
		    	second = parseInt(number % 60);

		    minute = minute >= 10 ? minute : "0" + minute;
		    second = second >= 10 ? second : "0" + second;
		    return minute + ":" + second;
    	},
    	//当前歌曲的总时间
    	TimeAll: function(){
    		return this.audio.duration;
    	}
    };
    module.exports = AudioFun;
});

src/page/index/main.js
define(function(require, exports, module){
    var audioFun = require('./Audio'),
    	audioDom = document.getElementById("myMusic"),
    	audio = new audioFun(audioDom);

    module.exports = {
    	init: function(){
    		var me = this;

    		me.MainControl = $("#MainControl");
    		me.SongName = $(".MusicList .List .Single .SongName");
    		me.LeftControl = $(".LeftControl");
    		me.RightControl = $(".RightControl");
    		me.VoiceEmp = $('.VoiceEmp');
    		me.VoiceFull = $('.VoiceFull');
    		me.Process = $('.Process');
    		me.ProcessYet = $('.ProcessYet');
    		me.VoiceProcess = $('.VoiceProcess');
    		me.VoiceProcessYet = $('.VoiceProcessYet');
    		me.ShowMusicList = $('.ShowMusicList');
    		me.MusicList = $('.MusicList');
    		me.MusicBox = $('.MusicBox');
    		me.BoxSongName = $(".MusicBox .ProcessControl .SongName");

    		me._initEvent();
    	},
    	_initEvent: function(){
    		var me = this;
    		me.MainControl.toggle(function(){
    			me._toggleStart($(this));
    		},function(){
    			me._toggleEnd($(this));
    		});
    		me.SongName.on('click', function(){
    			me._songListClick($(this));
    		});
    		me.LeftControl.on('click', function(){
    			me._leftControlClick($(this));
    		});
    		me.RightControl.on('click', function(){
    			me._rightControlClick($(this));
    		});
    		me.VoiceEmp.on('click', function(){
    			me._voiceEmpClick();
    		});
    		me.VoiceFull.on('click', function(){
    			me._voiceFullClick();
    		});
    		me.Process.on('click', function(e){
    			me._processClick(e, $(this));
    		});
    		me.ProcessYet.on('click', function(e){
    			me._processYetClick(e, $(this));
    		});
    		me.VoiceProcess.on('click', function(e){
    			me._voiceProcessClick(e, $(this));
    		});
    		me.VoiceProcessYet.on('click', function(e){
    			me._voiceProcessYetClick(e, $(this));
    		});
    		me.ShowMusicList.toggle(function(){
    			me._musicListToggleStart();
    		}, function(){
    			me._musicListToggleEnd();
    		});
    		console.log(audio);
    		//监听媒体文件结束的事件（ended），这边一首歌曲结束就读取下一首歌曲，实现播放上的循环播放
		    //audio.addEventListener('ended', me._audioListener, false);
    	},
    	_toggleStart: function(tar){
    		var me = this,
    			defaultsong = '';

    		tar.removeClass("MainControl").addClass("StopControl");
	        if (audio.src == "") {
	            defaultsong = $(".Single .SongName").eq(0).attr("KV");
	            me.BoxSongName.text(defaultsong);
	            $(".Single .SongName").eq(0).css("color", "#7A8093");
	            audio.src = "../../../static/Media/" + defaultsong + ".mp3";
	        }
	        audio.play();
	        me._timeSpan();
    	},
    	_toggleEnd: function(tar){
    		var me = this;
    		tar.removeClass("StopControl").addClass("MainControl");
        	audio.pause();
    	},
    	_songListClick: function(tar){
    		var me = this,
    			SongName = tar.attr("KV");

	        me.MainControl.removeClass("MainControl").addClass("StopControl");
	        tar.css("color", "#7A8093");
	        me.BoxSongName.text(SongName);

	        audio.src = "../../../static/Media/" + SongName + ".mp3";
	        audio.play();
	        me._timeSpan();
    	},
    	_leftControlClick: function(tar){
    		var me = this,
    			isTop,
    			prevSong;

    		me.SongName.each(function () {
	            if ($(this).css("color") == "rgb(122, 128, 147)") {
	                isTop = $(this).parent(".Single").prev(".Single").length == 0 ? true : false;  //检查是否是最顶端的歌曲
	                if (isTop) {
	                    prevSong = $(".Single").last().children(".SongName").attr("KV");
	                    $(".Single").last().children(".SongName").css("color", "#7A8093");
	                }
	                else {
	                    prevSong = $(this).parent(".Single").prev(".Single").children(".SongName").attr("KV");
	                    $(this).parent(".Single").prev(".Single").children(".SongName").css("color", "#7A8093");
	                }

	                audio.src = "../../../static/Media/" + prevSong + ".mp3";
	                me.BoxSongName.text(prevSong);
	                $(this).css("color", "#fff");
	                audio.play();
	                return false; 
	            }
	        });
    	},
    	_rightControlClick: function(tar){
    		var me = this,
    			isBottom,
    			nextSong;

    		me.SongName.each(function () {
	            if ($(this).css("color") == "rgb(122, 128, 147)") {
	                isBottom = $(this).parent(".Single").next(".Single").length == 0 ? true : false;  //检查是否是最尾端的歌曲

	                if (isBottom) {
	                    nextSong = $(".Single").first().children(".SongName").attr("KV");
	                    $(".Single").first().children(".SongName").css("color", "#7A8093");
	                }
	                else {
	                    nextSong = $(this).parent(".Single").next(".Single").children(".SongName").attr("KV");
	                    $(this).parent(".Single").next(".Single").children(".SongName").css("color", "#7A8093");
	                }

	                audio.src = "../../../static/Media/" + nextSong + ".mp3";
	                me.BoxSongName.text(nextSong);
	                $(this).css("color", "#fff");
	                audio.play();
	                return false; 
	            }
	        });
    	},
    	_voiceEmpClick: function(){
    		me.VoiceProcessYet.css("width", "0");
        	audio.volume = 0;
    	},
    	_voiceFullClick: function(){
    		me.VoiceProcessYet.css("width", "66px");
        	audio.volume = 1;
    	},
    	_processClick: function(e, tar){
    		var me = this,
    			processOffset,
    			processStart
    			processLength,
    			currentProces;

    		//播放进度条的基准参数
	        processOffset = tar.offset();
	        processStart = processOffset.left;
	        processLength = tar.width();


	        currentProces = e.clientX - processStart;
	        audio.DurationProcessRange(currentProces / processLength);
	        me.ProcessYet.css("width", currentProces);
    	},
    	_processYetClick: function(e, tar){
    		var me = this,
    			processOffset,
    			processStart,
    			processLength,
    			currentProces;

    		//播放进度条的基准参数
	        processOffset = me.Process.offset();
	        processStart = processOffset.left;
	        processLength = me.Process.width();

	        currentProces = e.clientX - processStart;
	        audio.DurationProcessRange(currentProces / processLength);
	        tar.css("width", currentProces);
    	},
    	_voiceProcessClick: function(e, tar){
    		var me = this,
    			voiceProcessOffset,
    			voiceProcessStart,
    			voiceProcessLength,
    			currentProces;

    		//音量进度条的基准参数
	        voiceProcessOffset = tar.offset();
	        voiceProcessStart = voiceProcessOffset.left;
	        voiceProcessLength = tar.width();

	        currentProces = e.clientX - voiceProcessStart;
	        audio.VolumeProcessRange(currentProces / voiceProcessLength);
	        me.VoiceProcessYet.css("width", currentProces);
    	},
    	_voiceProcessYetClick: function(){
    		var me = this,
    			voiceProcessOffset,
    			voiceProcessStart,
    			voiceProcessLength,
    			currentProces;

    		//音量进度条的基准参数
	        voiceProcessOffset = me.VoiceProcess.offset();
	        voiceProcessStart = voiceProcessOffset.left;
	        voiceProcessLength = me.VoiceProcess.width();

	        currentProces = e.clientX - voiceProcessStart;
	        audio.VolumeProcessRange(currentProces / voiceProcessLength);
	        tar.css("width", currentProces);
    	},
    	_musicListToggleStart: function(){
    		var me = this,
    			musicBoxRight,
    			musicBoxBottom;

    		me.MusicList.show();
	        musicBoxRight = me.MusicBox.offset().left + me.MusicBox.width();
	        musicBoxBottom = me.MusicBox.offset().top + me.MusicBox.height();

	        me.MusicList.css("left", musicBoxRight - me.MusicList.width());
	        me.MusicList.css("top", musicBoxBottom + 15);
    	},
    	_musicListToggleEnd: function(){
    		var me = this;

    		me.MusicList.hide();
    	},
    	_audioListener: function(){
    		var me = this,
    			isBottom,
    			nextSong;

    		me.SongName.each(function () {
	            if ($(this).css("color") == "rgb(122, 128, 147)") {
	                isBottom = $(this).parent(".Single").next(".Single").length == 0 ? true : false;  //检查是否是最尾端的歌曲

	                if (isBottom) {
	                    nextSong = $(".Single").first().children(".SongName").attr("KV");
	                    $(".Single").first().children(".SongName").css("color", "#7A8093");
	                }
	                else {
	                    nextSong = $(this).parent(".Single").next(".Single").children(".SongName").attr("KV");
	                    $(this).parent(".Single").next(".Single").children(".SongName").css("color", "#7A8093");
	                }

	                audio.src = "../../../static/Media/" + nextSong + ".mp3";
	                me.BoxSongName.text(nextSong);
	                $(this).css("color", "#fff");
	                audio.play();
	                return false; 
	            }
	        });
    	},
    	_timeSpan: function(){
		    var processYet = 0,
		    	currentTime = 0,
		    	timeAll = 0;

		    setInterval(function () {
		        processYet = (audio.currentTime / audio.duration) * 500;
		        me.ProcessYet.css("width", processYet);
		        vcurrentTime = audio.timeDispose(audio.currentTime);
		        timeAll = audio.timeDispose(audio.TimeAll());
		        $(".SongTime").html(currentTime + "&nbsp;|&nbsp;" + timeAll);
		    }, 1000);
    	}
    };
});